### PR TITLE
Adapt nix profile for fish.

### LINKED ISSUE #3980: Installation works in fish shell, but then it's not usable
Hi,

Installing nix from a fish shell (Archlinux), seems to work at first, but nix is then not usable.

**Steps To Reproduce**

1. Start an OS and use a fish shell (in my case: Archlinux)
2. Install nix with the official installation process: `curl -L https://nixos.org/nix/install | sh`
3. This will work and show the following message:

```
installing 'nix-2.3.7'
unpacking channels...

Installation finished!  To ensure that the necessary environment
variables are set, either log in again, or type

  . /home/vagrant/.nix-profile/etc/profile.d/nix.sh

in your shell.
```

4. Running `. /home/vagrant/.nix-profile/etc/profile.d/nix.sh` reveals it's not working, because fish is not sh. The following error message is given:

```
~/.nix-profile/etc/profile.d/nix.sh (line 6): Unsupported use of '='. In fish, please use 'set NIX_LINK $HOME/.nix-profile'.
from sourcing file ~/.nix-profile/etc/profile.d/nix.sh
        called on line 185 of file /usr/share/fish/config.fish
in function '.' with arguments '/home/user/.nix-profile/etc/profile.d/nix.sh'
source: Error while reading file “/home/user/.nix-profile/etc/profile.d/nix.sh”
```

**Expected behaviour**

The installation process should detect the default shell, and/or provide a profile for other shells (aka `nix.fish` inside `/home/user/.nix-profile/etc/profile.d/`)

### PR DESCRIPTION
**Pull request by Kineolyan**
This adapts the bash script to source to correctly setup nix in fish.

Fixes #3980 

_________________

Note that the adapted script is currently running on my Ubuntu 20.04 machine and succesfully helped me setup my environment to work on this :smile: 

I looked at the places where _nix-profile.sh_ was used. I think I correctly updated the makefiles, for both scripts and tests.
I read the test file for _nix-profile.sh_ but I am not sure to understand what happens in it, nor how the test is supposed to be run. I looked at the [hacking guide](https://hydra.nixos.org/build/140754926/download/1/manual/contributing/hacking.html) but I haven't found relevant instruction regarding the scripts. Can you help me for that, please :hand:

### CODE REVIEW

File: scripts/install-nix-from-closure.sh [hunk 1]
```diff
@@ -220,10 +220,10 @@ if [ -z "$NIX_INSTALLER_NO_CHANNEL_ADD" ]; then
     fi
 fi
 
-added=
-p=$HOME/.nix-profile/etc/profile.d/nix.sh
 if [ -z "$NIX_INSTALLER_NO_MODIFY_PROFILE" ]; then
+    # For bash-compatible shells: Bash, Zsh
```
Comment (SuperSandro2000) [inline scripts/install-nix-from-closure.sh:224]:
```suggestion
    # For bash-compatible shells like bash or zsh
```

File: scripts/install-nix-from-closure.sh [hunk 2]
```diff
@@ -246,6 +246,21 @@ if [ -z "$NIX_INSTALLER_NO_MODIFY_PROFILE" ]; then
             break
         fi
     done
+
+    # For Fish
+    # Make the shell source nix.fish during login.
```
Comment (SuperSandro2000) [inline scripts/install-nix-from-closure.sh:251]:
```suggestion
    # Make fish source nix.fish during login.
```

File: scripts/install-nix-from-closure.sh [hunk 3]
```diff
@@ -246,6 +246,21 @@ if [ -z "$NIX_INSTALLER_NO_MODIFY_PROFILE" ]; then
...
+    p=$HOME/.nix-profile/etc/profile.d/nix.fish
+    for i in .config/fish/config.fish; do
```
Comment (SuperSandro2000) [inline scripts/install-nix-from-closure.sh:253]: We don't need the for if the file can only be in one location.

File: scripts/nix-profile.fish.in [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        set -gx NIX_SSL_CERT_FILE "$NIX_LINK/etc/ca-bundle.crt"
+    end
+
+    set -gxp MANPATH "$NIX_LINK/share/man"
+
+    fish_add_path --prepend --global "$NIX_LINK/bin"
+    set -e NIX_LINK
+
+end
+
```
Comment (SuperSandro2000) [inline scripts/nix-profile.fish.in:33]: One final new line to much.

File: scripts/install-nix-from-closure.sh (hunk 3 above)
Comment (Kineolyan) [reply scripts/install-nix-from-closure.sh:252]: I agree. I wanted to reuse the same pattern for code consistency. I will simplify this.

**Comment by Kineolyan** [comment]
More questions about tests:
 - I don't see what is processing _tests/common.sh.in_. There is a `SHELL` variable that set with `@bash@` so I am assuming that we are changing it to bash and zsh.
 - If we really want to test it, we need to install fish on the test system. I have looked for a place referencing zsh because I am pretty sure that a system does not pack zsh by default. I did not have any luck with that approach.

**Comment by Kineolyan** [comment]
Sorry but after spending a week trying to understand how to setup a working test, I still don't have a clue.
I looked at the build system. I see some variable substitution, but I don't understand how it happens with _make_ magic.

I tried to call `nix-shell` directly in my test file to have fish in my test env, with something like:
```bash
nix-shell --pure -p fish --run '/usr/bin/fish -c '\''. /run/user/1000/nix-test/tests/nix-profile/nix-profile.fish'\'''
```
but without success. The reason was that the system did not know where and how to pick up fish.

I also tried having a dedicated nix expression for this and run
```bash
nix shell -f nix-profile-fish.nix -c '/usr/bin/fish -c '\''. /run/user/1000/nix-test/tests/nix-profile/nix-profile.fish'\'''
```
like in some other tests - _shell.sh_ among others.
But from what I understand in other tests, there are no actual dependencies being pulled. For example, one test is "using ruby", but this actually just a dummy file named _ruby_ calling `printf`.

So I would really appreciate any help on this matter.

At this point, I simply remove my shell test :(

**Comment by danielb2** [comment]
Can someone please help this maintainer? As a [fish](https://fishshell.com/) user, I'd love to see this

**Comment by Kineolyan** [comment]
Just in case, as the topic still interests some people, I updated the PR to be in sync with master.

**Comment by edolstra** [comment]
I guess we can close this now that #7014 has been merged?

**Comment by Kineolyan** [comment]
Absolutely @edolstra . I looked at the PR and I can see everything I put into this PR.
Thanks for finally adding this support :smiley:

**Pull request closed without merging**